DeepSpeech.py

```diff
@@ -79,32 +79,65 @@ def dense(name, x, units, dropout_rate=None, relu=True):
 
 
 def rnn_impl_lstmblockfusedcell(x, seq_length, previous_state, reuse):
-    # Forward direction cell:
-    fw_cell = tf.contrib.rnn.LSTMBlockFusedCell(Config.n_cell_dim, reuse=reuse)
+    with tf.variable_scope('cudnn_lstm/rnn/multi_rnn_cell/cell_0'):
+        fw_cell = tf.contrib.rnn.LSTMBlockFusedCell(Config.n_cell_dim,
+                                                    reuse=reuse,
+                                                    name='cudnn_compatible_lstm_cell')
 
-    output, output_state = fw_cell(inputs=x,
-                                   dtype=tf.float32,
-                                   sequence_length=seq_length,
-                                   initial_state=previous_state)
+        output, output_state = fw_cell(inputs=x,
+                                       dtype=tf.float32,
+                                       sequence_length=seq_length,
+                                       initial_state=previous_state)
 
     return output, output_state
 
 
-def rnn_impl_static_rnn(x, seq_length, previous_state, reuse):
+def rnn_impl_cudnn_rnn(x, seq_length, previous_state, _):
+    assert previous_state is None # 'Passing previous state not supported with CuDNN backend'
+
     # Forward direction cell:
-    fw_cell = tf.nn.rnn_cell.LSTMCell(Config.n_cell_dim, reuse=reuse)
+    if not rnn_impl_cudnn_rnn.cell:
+        with tf.variable_scope('rnn'):
+            fw_cell = tf.contrib.cudnn_rnn.CudnnLSTM(num_layers=1,
+                                                     num_units=Config.n_cell_dim,
+                                                     input_mode='linear_input',
+                                                     direction='unidirectional',
+                                                     dtype=tf.float32)
+        rnn_impl_cudnn_rnn.cell = fw_cell
+
+    output, output_state = rnn_impl_cudnn_rnn.cell(inputs=x,
+                                                   sequence_lengths=seq_length)
 
-    # Split rank N tensor into list of rank N-1 tensors
-    x = [x[l] for l in range(x.shape[0])]
+    return output, output_state
+
+# Hack: CudnnLSTM works similarly to Keras layers in that when you instantiate
+# the object it creates the variables, and then you just call it several times
+# to enable variable re-use. Because all of our code is structure in an old
+# school TensorFlow structure where you can just call tf.get_variable again with
+# reuse=True to reuse variables, we can't easily make use of the object oriented
+# way CudnnLSTM is implemented, so we save a singleton instance in the function,
+# emulating a static function variable.
+rnn_impl_cudnn_rnn.cell = None
```

Comment:
I get the explanation, maybe it is worth explaining where the solution from the hack comes from? Reference to some source code ?

Reply:
I don't understand. This comment was supposed to explain the solution. What's not clear? It doesn't come from anywhere, it's just a way to only instantiate the layer once.

Reply:
@reuben Maybe I just did not got the point of the comment then. The comment itself is clear, but I don't get the link between the comment and this code `.cell = None`.

Reply:
Ok, re-reading again, I get the point, it's playing along with `if not rnn_impl_cudnn_rnn.cell` above. I guess it's just the placement that distrubs me. Moving it before `def rnn_impl_cudnn_rnn` might make it more obvious? Or I just lack sleep.